crates/control/src/calibration_controller.rs

```diff
@@ -0,0 +1,285 @@
+        Parameter<Duration, "calibration_controller.initial_to_calibration_stabilization_delay">,
+    max_retries_per_capture: Parameter<u32, "calibration_controller.max_retries_per_capture">,
+
+    calibration_measurements: AdditionalOutput<Vec<Measurement>, "calibration_measurements">,
+    last_calibration_corrections:
+        AdditionalOutput<Option<Corrections>, "last_calibration_corrections">,
+}
+
+#[context]
+#[derive(Default)]
+pub struct MainOutputs {
+    pub calibration_command: MainOutput<Option<CalibrationCommand>>,
+}
+
+#[derive(Clone, Debug, Default, Serialize, Deserialize)]
+pub enum InternalCalibrationState {
+    #[default]
+    Inactive,
+    Initialize {
+        started_time: CycleTime,
+    },
+    LookAt {
+        target: Point2<Ground>,
+        camera: CameraPosition,
+        dispatch_time: CycleTime,
+    },
+    Capture {
+        camera: CameraPosition,
+        dispatch_time: CycleTime,
+    },
+    Process,
+    Finish,
+}
+
+impl From<InternalCalibrationState> for Option<CalibrationCommand> {
+    fn from(value: InternalCalibrationState) -> Self {
+        match value {
+            InternalCalibrationState::LookAt {
+                target,
+                camera,
+                dispatch_time,
+            } => Some(CalibrationCommand::LookAt {
+                target,
+                camera,
+                dispatch_time,
+            }),
+            InternalCalibrationState::Capture {
+                camera,
+                dispatch_time,
+            } => Some(CalibrationCommand::Capture {
+                camera,
+                dispatch_time,
+            }),
+            _ => None,
+        }
+    }
+}
+
+impl CalibrationController {
+    pub fn new(_context: CreationContext) -> Result<Self> {
+        Ok(Self {
+            current_calibration_state: InternalCalibrationState::default(),
+            look_at_list: generate_look_at_list(),
+            current_look_at_index: 0,
+            current_measurements: vec![],
+            last_calibration_corrections: None,
+            last_capture_retries: 0,
+        })
+    }
+
+    pub fn cycle(&mut self, mut context: CycleContext) -> Result<MainOutputs> {
+        if *context.primary_state != PrimaryState::Calibration {
+            self.current_calibration_state = InternalCalibrationState::Inactive;
+            return Ok(MainOutputs::default());
+        }
+
+        if let Some(new_state) = self.get_next_state(&context) {
+            self.current_calibration_state = new_state;
+        }
+
+        context
+            .calibration_measurements
+            .fill_if_subscribed(|| self.current_measurements.clone());
+        context
+            .last_calibration_corrections
+            .fill_if_subscribed(|| self.last_calibration_corrections);
+
+        let command: Option<CalibrationCommand> = self.current_calibration_state.clone().into();
+        Ok(MainOutputs {
+            calibration_command: command.into(),
+        })
```

Comment:
I don't like this clone().into() and .into() again, any ideas what can be done better?

Reply:
Replace the From implementation with a function that takes &self.
From moves the subject, so that'll get rid of one of the `into`s.
Having a well named function instead of "into" should also make it more clear what is happening and you don't have to write the type here anymore.

The one `into` for the main output is normal.

Reply:
Done, probably not the best function name, but should convey the message.